[Bug]: compilerOptions.paths does not work when using extends in tsconfig.json

### Version

1.42.1

### Steps to reproduce

1. `git clone git@github.com:alSergey/pw-reproduction.git`
2. `cd 3.10.3`
3. `npm ci`
4. `npx playwright test`

### Expected behavior

All tests passed successfully

### Actual behavior

```
Error: Cannot find module '~/playwright/const'
Require stack:
- /Users/seralekhin/Desktop/pw-reproduction/3.10.3/playwright/example.spec.ts
- /Users/seralekhin/Desktop/pw-reproduction/3.10.3/node_modules/playwright/lib/transform/transform.js
- /Users/seralekhin/Desktop/pw-reproduction/3.10.3/node_modules/playwright/lib/common/config.js
- /Users/seralekhin/Desktop/pw-reproduction/3.10.3/node_modules/playwright/lib/reporters/html.js
- /Users/seralekhin/Desktop/pw-reproduction/3.10.3/node_modules/playwright/lib/runner/reporters.js
- /Users/seralekhin/Desktop/pw-reproduction/3.10.3/node_modules/playwright/lib/runner/runner.js
- /Users/seralekhin/Desktop/pw-reproduction/3.10.3/node_modules/playwright/lib/program.js
- /Users/seralekhin/Desktop/pw-reproduction/3.10.3/node_modules/@playwright/test/cli.js

   at example.spec.ts:2

  1 | import { test, expect } from '@playwright/test';
> 2 | import { PW_URL } from '~/playwright/const';
    | ^
  3 |
  4 | test('has title', async ({ page }) => {
  5 |   await page.goto(PW_URL);

    at Object.<anonymous> (/Users/seralekhin/Desktop/pw-reproduction/3.10.3/playwright/example.spec.ts:2:1)
```

### Additional context

It seems to me that the problem is that you are not taking into account the tsconfig location and the baseUrl field. Here is a similar example, where tsconfig is generated differently and everything works.

1. `git clone git@github.com:alSergey/pw-reproduction.git`
2. `cd 3.5.3`
3. `npm ci`
4. `npx playwright test`

### Environment

```shell
System:
    OS: macOS 14.3.1
    CPU: (10) arm64 Apple M2 Pro
    Memory: 202.52 MB / 16.00 GB
  Binaries:
    Node: 16.20.2 - ~/.nvm/versions/node/v16.20.2/bin/node
    Yarn: 1.22.19 - ~/.nvm/versions/node/v16.20.2/bin/yarn
    npm: 8.19.4 - ~/.nvm/versions/node/v16.20.2/bin/npm
  IDEs:
    VSCode: 1.87.0 - /usr/local/bin/code
  Languages:
    Bash: 3.2.57 - /bin/bash
  npmPackages:
    @playwright/test: ^1.42.1 => 1.42.1
```

--- a/packages/playwright/src/transform/transform.ts
+++ b/packages/playwright/src/transform/transform.ts
@@ -30,7 +30,7 @@ import { getFromCompilationCache, currentFileDepsCollector, belongsToNodeModules
 const version = require('../../package.json').version;
 
 type ParsedTsConfigData = {
-  absoluteBaseUrl: string;
+  pathsBase?: string;
   paths: { key: string, values: string[] }[];
   allowJs: boolean;
 };
@@ -58,16 +58,15 @@ export function transformConfig(): TransformConfig {
 }
 
 function validateTsConfig(tsconfig: LoadedTsConfig): ParsedTsConfigData {
-  // Make 'baseUrl' absolute, because it is relative to the tsconfig.json, not to cwd.
   // When no explicit baseUrl is set, resolve paths relative to the tsconfig file.
   // See https://www.typescriptlang.org/tsconfig#paths
-  const absoluteBaseUrl = path.resolve(path.dirname(tsconfig.tsConfigPath), tsconfig.baseUrl ?? '.');
+  const pathsBase = tsconfig.absoluteBaseUrl ?? tsconfig.paths?.pathsBasePath;
   // Only add the catch-all mapping when baseUrl is specified
-  const pathsFallback = tsconfig.baseUrl ? [{ key: '*', values: ['*'] }] : [];
+  const pathsFallback = tsconfig.absoluteBaseUrl ? [{ key: '*', values: ['*'] }] : [];
   return {
     allowJs: !!tsconfig.allowJs,
-    absoluteBaseUrl,
-    paths: Object.entries(tsconfig.paths || {}).map(([key, values]) => ({ key, values })).concat(pathsFallback)
+    pathsBase,
+    paths: Object.entries(tsconfig.paths?.mapping || {}).map(([key, values]) => ({ key, values })).concat(pathsFallback)
   };
 }
 
@@ -132,7 +131,7 @@ export function resolveHook(filename: string, specifier: string): string | undef
         let candidate = value;
         if (value.includes('*'))
           candidate = candidate.replace('*', matchedPartOfSpecifier);
-        candidate = path.resolve(tsconfig.absoluteBaseUrl, candidate);
+        candidate = path.resolve(tsconfig.pathsBase!, candidate);
         const existing = resolveImportSpecifierExtension(candidate);
         if (existing) {
           longestPrefixLength = keyPrefix.length;

--- a/tests/playwright-test/resolver.spec.ts
+++ b/tests/playwright-test/resolver.spec.ts
@@ -505,6 +505,9 @@ test('should support extends in tsconfig.json', async ({ runInlineTest }) => {
     }`,
     'tsconfig.base1.json': `{
       "extends": "./tsconfig.base.json",
+      "compilerOptions": {
+        "allowJs": true,
+      },
     }`,
     'tsconfig.base2.json': `{
       "compilerOptions": {
@@ -518,7 +521,9 @@ test('should support extends in tsconfig.json', async ({ runInlineTest }) => {
         },
       },
     }`,
-    'a.test.ts': `
+    'a.test.js': `
+      // This js file is affected by tsconfig because allowJs is inherited.
+      // Next line resolve to the final baseUrl ("dir") + relative path mapping ("./foo/bar/util/*").
       const { foo } = require('util/file');
       import { test, expect } from '@playwright/test';
       test('test', ({}, testInfo) => {
@@ -526,6 +531,36 @@ test('should support extends in tsconfig.json', async ({ runInlineTest }) => {
       });
     `,
     'dir/foo/bar/util/file.ts': `
+      module.exports = { foo: 'foo' };
+    `,
+  });
+
+  expect(result.passed).toBe(1);
+  expect(result.exitCode).toBe(0);
+});
+
+test('should resolve paths relative to the originating config when extending and no baseUrl', async ({ runInlineTest }) => {
+  const result = await runInlineTest({
+    'tsconfig.json': `{
+      "extends": ["./dir/tsconfig.base.json"],
+    }`,
+    'dir/tsconfig.base.json': `{
+      "compilerOptions": {
+        "paths": {
+          "~/*": ["../mapped/*"],
+        },
+      },
+    }`,
+    'a.test.ts': `
+      // This resolves relative to the base tsconfig that defined path mapping,
+      // because there is no baseUrl in the final tsconfig.
+      const { foo } = require('~/file');
+      import { test, expect } from '@playwright/test';
+      test('test', ({}, testInfo) => {
+        expect(foo).toBe('foo');
+      });
+    `,
+    'mapped/file.ts': `
       module.exports = { foo: 'foo' };
     `,
   });

